internal/federationdomain/resolvedprovider/resolved_provider.go

```diff
@@ -1,30 +1,176 @@
+	// that the user's upstream group membership will be updated to make them belong to no
+	// upstream groups (before applying identity transformations to determine their downstream
+	// group memberships).
+	UpstreamGroups []string
+
+	// The portion of the user's session data which is specific to the upstream identity provider type.
+	// Refer to the fields of psession.CustomSessionData whose types are specific to an identity provider type.
+	// Set this to be the potentially updated IDP-specific session data. If no updates were required, then
+	// set this to nil.
+	IDPSpecificSessionData interface{}
+}
+
+// UpstreamAuthorizeRequestState is the state capturing the downstream authorization request, used as a parameter to
+// FederationDomainResolvedIdentityProvider.UpstreamAuthorizeRedirectURL to help formulate the upstream authorization
+// request. It includes the state param that should be sent in the upstream authorization request. It also includes
+// the information needed to create the PKCE and nonce parameters for the upstream authorization request. If the
+// upstream authorization request does not need PKCE and/or nonce params, then implementations of
+// FederationDomainResolvedIdentityProvider.UpstreamAuthorizeRedirectURL may choose to ignore those struct fields.
+type UpstreamAuthorizeRequestState struct {
+	EncodedStateParam string
+	PKCE              pkce.Code
+	Nonce             nonce.Nonce
+}
+
+type FederationDomainResolvedIdentityProvider interface {
+	// GetDisplayName returns the display name of this identity provider, as configured in the FederationDomain.
+	GetDisplayName() string
+
+	// GetProvider returns a representation of the upstream identity provider custom resource related to this
+	// identity provider for the FederationDomain, e.g. the OIDCIdentityProvider.
+	GetProvider() upstreamprovider.UpstreamIdentityProviderI
+
+	// GetSessionProviderType returns the type of session created by this identity provider.
+	GetSessionProviderType() psession.ProviderType
+
+	// GetIDPDiscoveryType returns the type of this identity provider, as shown by the IDP discovery endpoint.
+	GetIDPDiscoveryType() v1alpha1.IDPType
+
+	// GetIDPDiscoveryFlows returns the supported flows of this identity provider,
+	// as shown by the IDP discovery endpoint.
+	GetIDPDiscoveryFlows() []v1alpha1.IDPFlow
+
+	// GetTransforms returns the compiled version of the identity transformations and policies configured on the
+	// FederationDomain for this identity provider.
+	GetTransforms() *idtransform.TransformationPipeline
+
+	// CloneIDPSpecificSessionDataFromSession should reach into the provided session and return a clone
+	// of the field which is specific to the upstream identity provider type. If the session's field is
+	// nil, then return nil.
+	// Refer to the fields of psession.CustomSessionData whose types are specific to an identity provider type.
+	CloneIDPSpecificSessionDataFromSession(session *psession.CustomSessionData) interface{}
```

Comment:
Should this return an `Identity` instead of an `interface{}` since it seems a caller would have to cast this to an `Identity` anyway?

Reply:
That would be a reasonable alternative interface. However, I'm hoping to delete this function from the interface in a future commit, so I don't think we have to worry about it too much here.